source/fab/source_tree.py

```diff
@@ -27,42 +28,50 @@ def visit(self, candidate: Path) -> List[Path]:
         raise NotImplementedError("Abstract method must be implemented")
 
 
-class ExtensionVisitor(TreeVisitor):
+class SourceVisitor(TreeVisitor):
     def __init__(self,
-                 extension_map: Mapping[str, Union[Type[Task], Type[Command]]],
+                 source_map:
+                     List[Tuple[str, Union[Type[Task], Type[Command]]]],
                  command_flags_map: Mapping[Type[Command], List[str]],
                  state: SqliteStateDatabase,
                  workspace: Path,
                  task_handler: Callable):
-        self._extension_map = extension_map
+        self._source_map = source_map
         self._command_flags_map = command_flags_map
         self._state = state
         self._workspace = workspace
         self._task_handler = task_handler
 
     def visit(self, candidate: Path) -> List[Path]:
         new_candidates: List[Path] = []
-        try:
-            task_class = self._extension_map[candidate.suffix]
-            reader: TextReader = FileTextReader(candidate)
-
-            if issubclass(task_class, Analyser):
-                task: Task = task_class(reader, self._state)
-            elif issubclass(task_class, SingleFileCommand):
-                flags = self._command_flags_map.get(task_class, [])
-                task = CommandTask(
-                    task_class(Path(reader.filename), self._workspace, flags))
-            else:
-                message = \
-                    f"Unhandled class '{task_class}' in extension map."
-                raise TypeError(message)
-
-            self._task_handler(task)
-
-            new_candidates.extend(task.products)
-
-        except KeyError:
-            pass
+
+        task_class = None
+        for pattern, classname in self._source_map:
+            # Note we keep searching through the map
+            # even after a match is found; this means that
+            # later matches will override earlier ones
+            if match(pattern, str(candidate)):
+                task_class = classname
```

Comment:
I smell duplicate code. Is something wrong here?

Reply:
A candidate for the phase / map object as well I expect (it can have a `.select_class` method or something which does this)

Reply:
I put this into the new object I created in the end so the duplication is gone